### PR TITLE
Support non-tile-aligned widths in LayerNorm (interleaved), and use dest registers more effectively

### LINKED ISSUE #28365: Allow non-tiled aligned shapes for LayerNorm
See https://github.com/tenstorrent/tt-metal/issues/18548 and all of that issue's comments for details.

It was found that LayerNorm only handles widths that are tile aligned. Needs to handle all shapes.

Repro steps:

Run the following pytest.
```
from loguru import logger
import random
import pytest
import torch
import ttnn
import traceback

from tests.ttnn.utils_for_testing import assert_with_pcc
aten = torch.ops.aten

def test_layer_norm(device):
    dtype = ttnn.bfloat16
    dlayout = ttnn.TILE_LAYOUT
    epsilon = 1e-6

    # fail 196, 38
    # pass 32, 16, 20
    h = 1
    # fail 768
    # pass 32, 38
    w = 2 
    x = torch.zeros(1, h, w).to(torch.bfloat16)
    for i in range(h):
        for j in range(w):
            if (i + j) % 2 == 0:
                x[0,i,j] = 1
    weight = torch.ones([w]).to(torch.bfloat16)
    bias = torch.zeros([w]).to(torch.bfloat16)
    try:
        print(f'DATA TYPE {dtype} SHAPE {x.shape}')
        ref_value = aten.native_layer_norm.default(x, [w], weight, bias, epsilon)[0]
        
        tt_x = ttnn.from_torch(x, dtype=dtype, layout=dlayout, device=device)
        tt_weight = ttnn.from_torch(weight, dtype=dtype, layout=dlayout, device=device)
        tt_bias = ttnn.from_torch(bias, dtype=dtype, layout=dlayout, device=device)

        tt_result = ttnn.layer_norm(tt_x, epsilon=epsilon)
        tt_result = ttnn.to_torch(tt_result)
        
        inds = torch.where(ref_value != tt_result)
    except Exception as e:
        logger.warning(f"Test execution crashed: {e}")
        print(traceback.format_exc())
        raise e

    assert len(tt_result.shape) == len(ref_value.shape)
    assert tt_result.shape == ref_value.shape
    print(f"ref_value {ref_value}\n tt_result {tt_result}")
    assert_with_pcc(ref_value, tt_result, 0.999)
```

Expected Output:
Same as torch: -1, 1

Actual:
Something different: 5.5625, -0.1797

Proof of concept to get the correct results:

1. The 1 / W (padded) needs to be changed to 1 / non padded W
```
diff --git a/ttnn/cpp/ttnn/operations/normalization/layernorm/device/multi_core/layernorm_op_multi_core.cpp b/ttnn/cpp/ttnn/operations/normalization/layernorm/device/multi_core/layernorm_op_multi_core.cpp
index b4488a65f5..954e0587a7 100644
--- a/ttnn/cpp/ttnn/operations/normalization/layernorm/device/multi_core/layernorm_op_multi_core.cpp
+++ b/ttnn/cpp/ttnn/operations/normalization/layernorm/device/multi_core/layernorm_op_multi_core.cpp
@@ -449,7 +449,7 @@ operation::ProgramWithCallbacks layernorm_multi_core(
     }
 
     uint32_t curr_row = 0;
-    float winv = 1.0f / W;
+    float winv = 1.0f / a.logical_shape()[-1];
```

2. The reduce / matmul scaler needs to be limited to an arbitrary limit.
For matmul that means the code needs to be changed in the following way.
```
ttnn/cpp/ttnn/deprecated/tt_dnn/kernels/dataflow/generate_mm_scaler.hpp:FORCE_INLINE void generate_mm_scaler(const uint32_t cb_id, const uint32_t scaler) {

FORCE_INLINE void generate_mm_scaler(const uint32_t cb_id, const uint32_t scaler) {

    for (int i = 0; i < 128; i += 8) { // 16
        ptr[i] = single_packed_scalar;
    }
    for (int i = 256; i < 384; i += 8) { // 16
        ptr[i] = single_packed_scalar;
    }

FORCE_INLINE void generate_mm_scaler(const uint32_t cb_id, const uint32_t scaler, uint32_t n) {

    int face0_bound = (n < 16)? (n * 8) : 128;
    int face2_bound = (n < 16)? 256 : (256 + (n-16) * 8);

    for (int i = 0; i < face0_bound; i += 8) { // 16
        ptr[i] = single_packed_scalar;
    }       
    for (int i = 256; i < face2_bound; i += 8) { // 16
        ptr[i] = single_packed_scalar;
    }      
```
The diff is
```
--- a/ttnn/cpp/ttnn/deprecated/tt_dnn/kernels/dataflow/generate_mm_scaler.hpp
+++ b/ttnn/cpp/ttnn/deprecated/tt_dnn/kernels/dataflow/generate_mm_scaler.hpp
@@ -35,3 +35,37 @@ FORCE_INLINE void generate_mm_scaler(const uint32_t cb_id, const uint32_t scaler
 
     cb_push_back(cb_id, 1);
 }
+
+
+// Tile is assumed to have 16-bit elements
+// Scaler is assumed to be a 16-bit value double packed into a u32
+FORCE_INLINE void generate_mm_scaler(const uint32_t cb_id, const uint32_t scaler, const uint32_t n) {
+    cb_reserve_back(cb_id, 1);
+
+    constexpr uint32_t num_zeros_reads = 2048 / MEM_ZEROS_SIZE;
+    uint64_t zeros_noc_addr = get_noc_addr(MEM_ZEROS_BASE);
+    uint32_t write_addr = get_write_ptr(cb_id);
+    volatile tt_l1_ptr uint32_t* ptr = reinterpret_cast<volatile tt_l1_ptr uint32_t*>(write_addr);
+
+    // Fill tile with zeros
+    // TODO: src addr does not need to be rewritten. Update/add api for this
+    noc_async_read_one_packet_set_state(zeros_noc_addr, MEM_ZEROS_SIZE);
+    for (uint32_t i = 0; i < num_zeros_reads; ++i) {
+        noc_async_read_one_packet_with_state(zeros_noc_addr, write_addr);
+        write_addr += MEM_ZEROS_SIZE;
+    }
+    noc_async_read_barrier();
+
+    int face0_bound = (n < 16)? (n * 8) : 128;
+    int face2_bound = (n < 16)? 256 : (256 + (n-16) * 8);
+
+    uint32_t single_packed_scalar = scaler & 0xFFFF;
+    for (int i = 0; i < face0_bound; i += 8) {
+        ptr[i] = single_packed_scalar;
+    }
+    for (int i = 256; i < face2_bound; i += 8) {
+        ptr[i] = single_packed_scalar;
+    }
+
+    cb_push_back(cb_id, 1);
+}
```
For reduce, the code needs to be
```
        int outer_bound = (n < 16)? 1 : 2;
        int inner_bound = (n < 16)? n/2 : 8;

        for (int k = 0; k < outer_bound; ++k) {
            uint32_t idx = k << 7; // * 128
            for (int j = 0; j < inner_bound; ++j) {
                ptr[idx + j] = scaler;
            }
            inner_bound = (n - 16)/2;
        }
```
The diff is
```
--- a/ttnn/cpp/ttnn/deprecated/tt_dnn/kernels/dataflow/generate_reduce_scaler.hpp
+++ b/ttnn/cpp/ttnn/deprecated/tt_dnn/kernels/dataflow/generate_reduce_scaler.hpp
@@ -36,6 +36,38 @@ FORCE_INLINE void generate_reduce_scaler(const uint32_t cb_id, const uint32_t sc
 
     cb_push_back(cb_id, 1);
 }
+FORCE_INLINE void generate_reduce_scaler(const uint32_t cb_id, const uint32_t scaler, const uint32_t n) {
+    cb_reserve_back(cb_id, 1);
+
+    constexpr uint32_t num_zeros_reads = 2048 / MEM_ZEROS_SIZE;
+    uint64_t zeros_noc_addr = get_noc_addr(MEM_ZEROS_BASE);
+    uint32_t write_addr = get_write_ptr(cb_id);
+    volatile tt_l1_ptr uint32_t* ptr = reinterpret_cast<volatile tt_l1_ptr uint32_t*>(write_addr);
+
+    // Fill tile with zeros
+    // TODO: src addr does not need to be rewritten. Update/add api for this
+    noc_async_read_one_packet_set_state(zeros_noc_addr, MEM_ZEROS_SIZE);
+    for (uint32_t i = 0; i < num_zeros_reads; ++i) {
+        noc_async_read_one_packet_with_state(zeros_noc_addr, write_addr);
+        write_addr += MEM_ZEROS_SIZE;
+    }
+    noc_async_read_barrier();
+
+    if (scaler != 0) {
+        int outer_bound = (n < 16)? 1 : 2;
+        int inner_bound = (n < 16)? n/2 : 8;
+
+        for (int k = 0; k < outer_bound; ++k) {
+            uint32_t idx = k << 7; // * 128
+            for (int j = 0; j < inner_bound; ++j) {
+                ptr[idx + j] = scaler;
+            }
+            inner_bound = (n - 16)/2;
+        }
+    }
+
+    cb_push_back(cb_id, 1);
+}
```
I just hard coded the limits to 2

ttnn/cpp/ttnn/operations/normalization/layernorm/device/kernels/dataflow/reader_unary_interleaved_ln.cpp
```
        generate_mm_scaler(cb_in_2, scaler, 2);
#else
        generate_reduce_scaler(cb_in_2, scaler, 2);
```

This issue will focus on the reduce only. A separate issue will need to be created to look at matmul as the last task of this issue.

### PR DESCRIPTION
**Pull request by rmillerTT**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/28365

### Problem description
This PR only impacts non-sharded and non-distributed LayerNorm. Previously, LN was using the padded tensor widths in the program factory, meaning that the reduce scalar (1/W) was incorrect. Further, the variance (for non-Welford LN) was incorrect, as the columns beyond the tensor width were being accumulated. This lead to a periodicity in the error that aligned with tile boundaries, as seen below.

This PR also sets the block size (the number of tiles processed into dest regs) to 4 (if fp32 accumulation is enabled) or 8 (otherwise) always. Previously, the block size was depdendent on the tensor width, leading to many widths processing only 1, 2, or 3 tiles at a time.

This PR addresses these issues to improve both accuracy and average runtime (see below).

Relative error before 
(fa_rand() input distribution over various tensor widths sampled at intervals of 27 to make it not tile aligned):
bfloat16:
image

float32:
image

Relative error now:
bfloat16:
image

float32:
image

These changes also improved the runtime of many cases, some substantially. **The geometric mean speedup of all unit tests in test_layer_norm.py was 7.02%**. Multiple single-core cases saw runtime improvements over 50%, as shown below.

Top 10 speedups:
image

Top 10 slowdowns:
image

### What's changed
Summarized above

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes https://github.com/tenstorrent/tt-metal/actions/runs/20467503549
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable) https://github.com/tenstorrent/tt-metal/actions/runs/20467508588
- [x] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable) https://github.com/tenstorrent/tt-metal/actions/runs/20467513296
- [x] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable) https://github.com/tenstorrent/tt-metal/actions/runs/20467518469
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable)
- [ ] [Galaxy quick](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-quick.yaml) CI passes (if applicable)
- [ ] [Galaxy demo tests, for Llama](https://github.com/tenstorrent/tt-metal/actions/workflows/galaxy-demo-tests.yaml) CI passes, if applicable, because of current Llama work
- [ ] (For runtime and ops writers) [T3000 unit tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-unit-tests.yaml) CI passes (if applicable, since this is run on push to main)
- [ ] (For models and ops writers) [T3000 demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/t3000-demo-tests.yaml) CI passes (if applicable, since this is required for release)
- [ ] [cpp-unit-tests](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml) job passes
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -332,7 +332,7 @@ def run(self) -> None:
             "ttnn/operations/data_movement/**/*",
             "ttnn/operations/moreh/**/*",
             "ttnn/kernel/*",
-            "ttnn/operations/normalization/kernel_util/compute/*",
+            "ttnn/operations/normalization/kernel_util/*",
```
Comment (bbradelTT) [inline setup.py]: Should this be `ttnn/operations/normalization/kernel_util/**/*` ?

File: tests/ttnn/unit_tests/operations/fused/test_layer_norm.py [hunk 2]
```diff
@@ -402,3 +402,43 @@ def test_layer_norm_across_dtypes(*, device: ttnn.Device, dim_a: int, dim_b: int
...
+@pytest.mark.parametrize("dtype", [torch.bfloat16, torch.float32])
+def test_layer_norm_with_padding(device, use_welford, dtype):
+    """
+    Test layer norm on a tensor that is padded with zeros
+    in the width dimension.
+    Compare against analytic layer norm calculation: (x - mean) / sqrt(var + eps)
+    Only tests Welford layernorm, since legacy reduce doesn't give the correct
+    result for partially-filled tiles.
+    """
+    # TODO Test height padding
```
Comment (bbradelTT) [inline tests/ttnn/unit_tests/operations/fused/test_layer_norm.py]: Please do this in this PR.

File: ttnn/cpp/ttnn/operations/normalization/layernorm/device/layernorm_op_multi_core.cpp [hunk 3]
```diff
@@ -237,8 +229,8 @@ LayerNormMultiCoreProgramFactory::cached_program_t LayerNormMultiCoreProgramFact
     uint32_t im2_t = 2;  //
 
     bool large_tensor_needed = false;
-    constexpr uint32_t no_weights_max_size = 120;
-    constexpr uint32_t with_weights_max_size = 60;
+    constexpr uint32_t with_weights_max_size = 56;
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/normalization/layernorm/device/layernorm_op_multi_core.cpp:238]: Please add a comment about how this was determined empirically.

File: ttnn/cpp/ttnn/operations/normalization/layernorm/device/layernorm_op_multi_core.cpp [hunk 4]
```diff
@@ -543,22 +520,22 @@ LayerNormMultiCoreProgramFactory::cached_program_t LayerNormMultiCoreProgramFact
...
-             beta_dram_addr,
-             b_dram_addr}  // 6-8
-        );
+        std::vector<uint32_t> reader_runtime_args = {
+            a_addr,
+            num_tile_rows_per_core,
+            Wt,
+            tile_offset,
+            packed_one_value,
+            e.u,  // 0-5
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/normalization/layernorm/device/layernorm_op_multi_core.cpp]: Please make the comment easier to understand.

Also, should the union not be used?

File: ttnn/cpp/ttnn/operations/normalization/kernel_util/generic/blocked_range.h [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        SizeT start() const { return start_; }
+        SizeT last() const { return end_; }
+        SizeT remainder() const { return block_size_ - size(); }
+        SizeT full_block_size() const { return block_size_; }
+
+        GlobalRange global() const { return GlobalRange{start_, end_}; }
+        LocalRange local() const { return LocalRange{size()}; }
+
+        SizeT to_global(SizeT local) const { return start_ + local; }
+        SizeT to_local(SizeT global) const { return global - start_; }
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/normalization/kernel_util/generic/blocked_range.h]: Is this used anywhere?
If not, consider removing. One issue is that the resulting local may not be in this block.

File: ttnn/cpp/ttnn/operations/normalization/kernel_util/dataflow/custom_tiles.h [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    const uint16_t scaler_uint16 = scaler >> 16;
+
+    constexpr uint32_t num_zeros_reads = 2048 / MEM_ZEROS_SIZE;
+    static_assert(num_zeros_reads > 0, "num_zeros_reads must be greater than 0");
+    uint64_t zeros_noc_addr = get_noc_addr(MEM_ZEROS_BASE);
+    uint32_t write_addr = get_write_ptr(cb_id);
+    volatile tt_l1_ptr uint16_t* ptr = reinterpret_cast<volatile tt_l1_ptr uint16_t*>(write_addr);
+
+    // Fill tile with zeros
+    // TODO: src addr does not need to be rewritten. Update/add api for this
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/normalization/kernel_util/dataflow/custom_tiles.h]: Please either create an issue or fix this in this PR.

File: ttnn/cpp/ttnn/operations/normalization/kernel_util/compute/numeric.h [hunk 7]
```diff
@@ -137,16 +147,18 @@ inline void row_wise_accumulate_with_epilogue(
     uint32_t cb_out,
     uint32_t num_tiles,
     uint32_t block_size,
+    uint32_t N,
     Epilogue epilogue = detail::no_op,
     AdditionalCBs... additional_cbs) {
-    constexpr bool pop_input = pop_input_policy == policies::PopInputPolicy::POP;
     constexpr bool wait_at_end = wait_at_end_policy == policies::WaitAtEndPolicy::WAIT;
-
+    const auto last_tile_partial = N % tt::constants::TILE_WIDTH > 0;
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/normalization/kernel_util/compute/numeric.h:158]: Please add a comment with an explanation.

**Review by bbradelTT — COMMENTED** [review]
Looks good. Will approve once pipelines are run.

**Review by vsureshTT — APPROVED** [review]

**Comment by rmillerTT** [comment]
L2 nightly passing: https://github.com/tenstorrent/tt-metal/actions/runs/20467536740

**Review by nsextonTT — APPROVED** [review]
🔥
